python_modules/libraries/dagster-airlift/dagster_airlift/core/components/airflow_instance/component.py

```diff
@@ -93,7 +143,67 @@ def _get_instance(self) -> dg_airlift_core.AirflowInstance:
     def build_defs(self, context: ComponentLoadContext) -> Definitions:
         defs = build_job_based_airflow_defs(
             airflow_instance=self._get_instance(),
+            mapped_defs=apply_mappings(defs_from_subdirs(context), self.mappings),
         )
         for post_processor in self.asset_post_processors or []:
             defs = post_processor(defs)
         return defs
+
+
+EXCLUDED_DIRS = ["__pycache__", ".tox", ".venv"]
+
+
+def defs_from_subdirs(context: ComponentLoadContext) -> Definitions:
+    """Get all definitions coming from subdirectories in the current context. We explicitly don't want to include the current directory in the search, to avoid including the current component."""
+    return DefsFolderComponent(
+        path=context.path,
+        children=find_components_from_context(context, EXCLUDED_DIRS),
+        asset_post_processors=None,
+    ).build_defs(context)
+
+
+def handle_iterator(
+    mappings: Optional[Sequence[AirflowDagMapping]],
+) -> Iterator[
+    tuple[Union[TaskHandle, DagHandle], Sequence[Union[InAirflowAsset, InDagsterAssetRef]]]
+]:
+    if mappings is None:
+        return
+    for mapping in mappings:
+        for task_mapping in mapping.task_mappings:
+            yield (
+                TaskHandle(dag_id=mapping.dag_id, task_id=task_mapping.task_id),
+                task_mapping.assets,
+            )
+        yield DagHandle(dag_id=mapping.dag_id), mapping.assets
+
+
+def apply_mappings(defs: Definitions, mappings: Sequence[AirflowDagMapping]) -> Definitions:
+    specs = {spec.key: spec for spec in defs.get_all_asset_specs()}
+    key_to_handle_mapping = {}
+    additional_assets = []
+
+    for handle, asset_refs in handle_iterator(mappings):
+        if not asset_refs:
+            continue
+        for asset in asset_refs:
+            if isinstance(asset, AssetKey):
+                if asset not in specs:
+                    raise ValueError(f"Asset with key {asset} not found in definitions")
```

Comment:
nit: maybe DagsterInvalidDefinitionError would be a bit more specific here?

Reply:
fine with that